src/service/ComposeDocument.ts

```diff
@@ -150,6 +83,43 @@ export class ComposeDocument {
         this.documentSettings = params;
     }
 
+    public async getPositionInfo(params: ExtendedPositionParams): Promise<PositionInfo> {

```

Comment:
I am going to trust you on the logic of `getPositionInfo` and `getFirstLinePositionInfo`, but maybe consider writing a comment for these two functions describing in high-level terms what they do. In particular, I suspect the order in which the regular expressions are tried is significant.

Reply:
Good idea. I also agree with your overview comment for writing tests.